Add dry run functionality to migrate virctl command

fossedihelm opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. Follow the instructions for writing a release note from k8s: https://git.k8s.io/community/contributors/guide/release-notes.md
-->

**What this PR does / why we need it**:
Allows user to use --dry-run flag for migrate virtctl command.
Passing this flag to the migrate command will do all checks showing possible failures, but changes will not be effective.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
added DryRun mode to virtcl to migrate command
migrate requests now have a body with MigrateOptions
```

iholder-redhat commented on pkg/virtctl/vm/vm.go:
```diff
@@ -127,6 +127,7 @@ func NewMigrateCommand(clientConfig clientcmd.ClientConfig) *cobra.Command {
 			return c.Run(args)
 		},
 	}
+	cmd.Flags().BoolVar(&dryRun, "dry-run", false, "--dry-run=false: Flag used to set whether to perform a dry run or not. If true the command will be executed without performing any changes.")
```
I see this string is (`""--dry-run=false: Fla..."`) is duplicated many times in this file, can we save it as a const and re-use it?

fossedihelm replied:
Totally agree! I noticed and reworked it in the current task I'm working. I will put that here. Thanks

iholder-redhat commented on staging/src/kubevirt.io/client-go/apis/core/v1/types.go:
```diff
@@ -1890,6 +1890,21 @@ type StopOptions struct {
+	// - All: all dry run stages will be processed
+	// +optional
+	// +listType=atomic
+	DryRun []string `json:"dryRun,omitempty" protobuf:"bytes,1,rep,name=dryRun"`
```
Can you explain why this needs to be of an array type? why not a regular string?

fossedihelm replied:
Thanks. Sure, I used the same format used by [kubernetes](https://github.com/kubernetes/apimachinery/blob/master/pkg/apis/meta/v1/types.go#L529), following the same rules. This, imo, prevent the need to wrap the single string into a slice when propagating it toward k8s APIs, and also helps in case of future developments. Let me know if you disagree, we can talk about it :)

iholder-redhat replied:
I wonder why Kubernetes defined it this way..
Anyways you're right :) thanks

fossedihelm replied:
Just a guess: from the description seems that maybe in the future, you can specify in which stages perform a dry-run. :) Thanks

iholder-redhat commented on tests/vm_test.go:
```diff
@@ -1220,6 +1220,30 @@ var _ = Describe("[rfe_id:1177][crit:medium][vendor:cnv-qe@redhat.com][level:com
+					migrateCommand := tests.NewRepeatableVirtctlCommand(vm.COMMAND_MIGRATE, "--dry-run", "--namespace", virtualMachine.Namespace, virtualMachine.Name)
+					Expect(migrateCommand()).To(Succeed())
+
+					By("Check that no migration was actually created")
+					_, err = virtClient.VirtualMachineInstanceMigration(virtualMachine.Namespace).Get(virtualMachine.Name, &metav1.GetOptions{})
+					Expect(errors.IsNotFound(err)).To(BeTrue())
```
When we want to validate that migration **is** created we usually use an `Eventually` clause since it might take some time for it to actually be created. Here's an example:
```
Eventually(func() error {
	_, err := virtClient.VirtualMachineInstance(updatedVM.Namespace).Get(updatedVM.Name, &k8smetav1.GetOptions{})
	return err
}, 300*time.Second, 1*time.Second).Should(Succeed())
```

Maybe we can do something similar here? Maybe something like:
```
Consistently(func() error {
        _, err = virtClient.VirtualMachineInstanceMigration(virtualMachine.Namespace).Get(virtualMachine.Name, &metav1.GetOptions{})
	return errors.IsNotFound(err)
}, 60*time.Second, 5*time.Second).Should(BeTrue(), "migration should not be created in a dry run mode")
```

fossedihelm replied:
Thanks for the suggestion. I will fix it. :)

iholder-redhat reviewed (commented):
Thanks @fossedihelm! Looks good :)
See comments below

fossedihelm commented on pkg/virtctl/vm/vm_test.go:
```diff
@@ -158,15 +158,18 @@ var _ = Describe("VirtualMachine", func() {
 
 			kubecli.MockKubevirtClientInstance.EXPECT().VirtualMachine(k8smetav1.NamespaceDefault).Return(vmInterface).Times(1)
-			vmInterface.EXPECT().Migrate(vm.Name).Return(nil).Times(1)
+			vmInterface.EXPECT().Migrate(vm.Name, &v1.MigrateOptions{}).Return(nil).Times(1)
```
Oh yes, thanks so much. I've missed it. Will fix

iholder-redhat commented:
Thank you @fossedihelm! Great work
/lgtm

fossedihelm commented:
@iholder-redhat I force-pushed due to a conflict :)

iholder-redhat commented:
> @iholder-redhat I force-pushed due to a conflict :)

:+1: 
/ok-to-test
/lgtm

fossedihelm commented:
/retest

fossedihelm commented:
/retest

iholder-redhat commented:
/lgtm

fossedihelm commented:
@iholder-redhat Just a push to resolve conflicts, no other changes :)

fossedihelm commented:
/retest

iholder-redhat commented:
> @iholder-redhat Just a push to resolve conflicts, no other changes :)

Thank you!
/lgtm

iholder-redhat commented:
once again :)
/lgtm

fossedihelm commented:
/retest

fossedihelm commented:
/retest

iholder-redhat commented:
/retest
/lgtm

xpivarc commented:
I think it would be great if we can add mock tests in the `subresource_test.go`. It would basically allow us to prepare expectations and reactors that will verify all client calls will have the dryrun option passed down.

enp0s3 reviewed (approved):
@fossedihelm Thank you!
/approve

enp0s3 commented:
/hold
@xpivarc Sorry I didn't see your comment, please feel free to unhold when coversation with @fossedihelm on this subject will be resolved. Thanks!

fossedihelm commented:
> I think it would be great if we can add mock tests in the `subresource_test.go`. It would basically allow us to prepare expectations and reactors that will verify all client calls will have the dryrun option passed down. Reducing the risk of introducing side effects in the future.

@xpivarc I added the mock virctl in the `subresource_test.go` file and rebased due to conflicts. Tell me what do you think about the implementation, thanks :slightly_smiling_face: 
@enp0s3 Thanks

iholder-redhat commented on pkg/virt-api/rest/subresource_test.go:
```diff
@@ -99,6 +115,19 @@ var _ = Describe("VirtualMachineInstance Subresources", func() {
+
+		virtClient.EXPECT().CoreV1().Return(kubeClient.CoreV1()).AnyTimes()
+		virtClient.EXPECT().VirtualMachine(k8smetav1.NamespaceDefault).Return(vmClient).AnyTimes()
+		virtClient.EXPECT().VirtualMachine("").Return(vmClient).AnyTimes()
```
Is this a leftover?
Referring to `virtClient.EXPECT().VirtualMachine("").Return(vmClient).AnyTimes()`.

fossedihelm replied:
Thanks @iholder-redhat some tests (i.e. [this one](https://github.com/fossedihelm/kubevirt/blob/97ce2f4ee062f5cc211efbd129080b483cfa67a4/pkg/virt-api/rest/subresource_test.go#L301)) check the failure if the namespace is not passed down, for that reason I added this line. If there is another way to manage it, please give me an advice :slightly_smiling_face:

iholder-redhat commented on pkg/virt-api/rest/subresource_test.go:
```diff
@@ -99,6 +115,19 @@ var _ = Describe("VirtualMachineInstance Subresources", func() {
+		virtClient.EXPECT().VirtualMachine(k8smetav1.NamespaceDefault).Return(vmClient).AnyTimes()
+		virtClient.EXPECT().VirtualMachine("").Return(vmClient).AnyTimes()
+		virtClient.EXPECT().VirtualMachineInstance(k8smetav1.NamespaceDefault).Return(vmiClient).AnyTimes()
+		virtClient.EXPECT().VirtualMachineInstance("").Return(vmiClient).AnyTimes()
```
same as above.
Why would you expect someone asking for a VMI from an empty namespace?

iholder-redhat commented on pkg/virt-api/rest/subresource_test.go:
```diff
@@ -109,7 +138,8 @@ var _ = Describe("VirtualMachineInstance Subresources", func() {
 		flag.Set("kubeconfig", "")
 		flag.Set("master", server.URL())
-		app.virtCli, _ = kubecli.GetKubevirtClientFromFlags(server.URL(), "")
+		//app.virtCli, _ = kubecli.GetKubevirtClientFromFlags(server.URL(), "")
```
leftover

fossedihelm replied:
Thanks, I will fix that

xpivarc commented on pkg/virt-api/rest/subresource_test.go:
```diff
@@ -1525,30 +1218,15 @@ var _ = Describe("VirtualMachineInstance Subresources", func() {
+			vmClient.EXPECT().Get(testVMName, &k8smetav1.GetOptions{}).Return(&vm, nil)
+			vmiClient.EXPECT().Get(testVMName, &k8smetav1.GetOptions{}).Return(&vmi, nil)
 
+			migrateClient.EXPECT().Create(gomock.Any(), gomock.Any()).Do(
```
Perfect!

xpivarc reviewed (approved):
Initially, I wanted only to add a test that would use the mock approach but I think it makes sense to rewrite it all(If there are no objections).

iholder-redhat commented on pkg/virt-api/rest/subresource_test.go:
```diff
@@ -1411,54 +1139,48 @@ var _ = Describe("VirtualMachineInstance Subresources", func() {
-			close(done)
-		}, 5)
+		},
+			table.Entry("", &v1.MigrateOptions{}),
```
Description is left empty here. I guess something like `"In a real run"` or `"without dry-run option"` will do.

IMHO It's better to not keep these empty for two reasons, first being that it's clearer, and secondly since if another Entry would show up with an empty description it would create a mess.

P.S. - you have more of Entries with no description in this file, I'm not going to comment the same for every one but the same thing applies to all :)

fossedihelm replied:
Thanks so much. Actually you're totally right. I will fix that :slightly_smiling_face:

iholder-redhat reviewed (commented):
Thanks @fossedihelm !

fossedihelm commented:
@iholder-redhat I pushed the fixes
@xpivarc Added test to check if dryrun opt is propagated in start stop and restart virctl command
Thanks to both of you!

iholder-redhat commented on pkg/virt-api/rest/subresource_test.go:
```diff
@@ -513,8 +509,10 @@ var _ = Describe("VirtualMachineInstance Subresources", func() {
-				close(done)
-			})
+			},
+				table.Entry("with default", &v1.RestartOptions{GracePeriodSeconds: &gracePeriodZero}),
```
That's better than what I came up with :) :+1:

iholder-redhat reviewed (commented):
Thanks @fossedihelm!
/lgtm

@xpivarc feel free to cancel hold if you have no more feedback

fossedihelm commented:
/retest

xpivarc reviewed (commented):
/hold cancel

fossedihelm commented:
/retest
